mlbench/refimpls/pytorch/utils/log.py

```diff
@@ -31,17 +31,59 @@ def debug(content, who='all'):
 
 def warning(content, who='all'):
     if who == 'all' or who == dist.get_rank():
-        logger.warning("\033[0;31m{}\033[0m".format(content))
+        logger.warning("{}".format(content))
 
 
 def critical(content, who='all'):
     if who == 'all' or who == dist.get_rank():
-        logger.critical("\033[0;104m{}\033[0m".format(content))
+        logger.critical("{}".format(content))
+
+
+def post_metrics(payload):
+    """Post information via kubernetes client.
+
+    Example:
+
+        payload = {
+            "run_id": "1",
+            "name": "accuracy",
+            "cumulative": False,
+            "date": "2018-08-14T09:21:44.331823Z",
+            "value": "1.0",
+            "metadata": "some additional data"
+        }
+
+    See `KubeMetric` in kubemetric.py for the fields and types.
+
+    """
+    from kubernetes import config, client
+    import requests
+    import json
+
+    config.load_incluster_config()
+    configuration = client.Configuration()
+    api_instance = client.CoreV1Api(client.ApiClient(configuration))
+
+    # For the moment, the namespace and label_selectors are hard-coded.
+    namespace = 'default'
+    label_selector = 'component=master,app=mlbench'
+    try:
+        api_response = api_instance.list_namespaced_pod(
+            namespace, label_selector=label_selector)
+    except ApiException as e:
+        print("Exception when calling CoreV1Api->list_namespaced_pod: %s\n" % e)
+
+    assert len(api_response.items) == 1
+    ip = api_response.items[0].status.pod_ip
+
+    response = requests.post("http://{ip}/api/metrics/".format(ip=ip), data=payload)
```

Comment:
Is this blocking? Did you try it by wrapping the call with the `kubernetes.stream` function?

Reply:
I was thinking about adding non-blocking version later. But it will not take too long. I can update this module before merge.

Reply:
`kubernetes.stream` is used for watching api endpoints, no? What we want is more like having another thread to `post`?

Reply:
It is intended to watch api endpoints, yes. But when I looked into the source, it looks like it can wrap any api call and it will do the call using websockets without blocking instead of waiting for a call to finish.

the client also has a (mostly undocumented) functionality for threads already, with the async=True keyword-argument for all. E.g. https://github.com/kubernetes-client/gen/pull/67